copulas/__init__.py

```diff
@@ -21,19 +21,29 @@ class NotFittedError(Exception):
 
 
 @contextlib.contextmanager
-def random_seed(seed):
-    """Context manager for managing the random seed.
+def set_random_state(random_state, set_model_random_state):
+    """Context manager for managing the random state.
 
     Args:
-        seed (int):
-            The random seed.
+        random_state (int or np.random.RandomState):
+            The random seed or RandomState.
+        set_model_random_state (function):
+            Function to set the random state on the model.
     """
-    state = np.random.get_state()
-    np.random.seed(seed)
+    original_state = np.random.get_state()
+
+    if isinstance(random_state, int):
+        desired_state = np.random.RandomState(seed=random_state)
+    else:
+        desired_state = random_state
+
+    np.random.set_state(desired_state.get_state())
+
     try:
         yield
     finally:
-        np.random.set_state(state)
+        set_model_random_state(desired_state)
```

Comment:
We should not be caputring the `desired_state`, but rather the `current state of numpy` as returned by `np.random.get_state()`